### ISSUE TITLE
[Feature Request]:  GROUP BY WITH ROLLUP

### ISSUE THREAD

**Issue by dengn** [matrixorigin/matrixone#15861]
### Is there an existing issue for the same feature request?

- [X] I have checked the existing issues.

### Describe the feature you'd like

The GROUP BY clause permits a WITH ROLLUP modifier that causes summary output to include extra rows that represent higher-level (that is, super-aggregate) summary operations. ROLLUP thus enables you to answer questions at multiple levels of analysis with a single query. For example, ROLLUP can be used to provide support for OLAP (Online Analytical Processing) operations.

CREATE TABLE sales
(
    year    INT,
    country VARCHAR(20),
    product VARCHAR(32),
    profit  INT
);

mysql> SELECT year, SUM(profit) AS profit
       FROM sales
       GROUP BY year;
+------+--------+
| year | profit |
+------+--------+
| 2000 |   4525 |
| 2001 |   3010 |
+------+--------+

The output shows the total (aggregate) profit for each year. To also determine the total profit summed over all years, you must add up the individual values yourself or run an additional query. Or you can use ROLLUP, which provides both levels of analysis with a single query. Adding a WITH ROLLUP modifier to the GROUP BY clause causes the query to produce another (super-aggregate) row that shows the grand total over all year values:

mysql> SELECT year, SUM(profit) AS profit
       FROM sales
       GROUP BY year WITH ROLLUP;
+------+--------+
| year | profit |
+------+--------+
| 2000 |   4525 |
| 2001 |   3010 |
| NULL |   7535 |
+------+--------+

### Describe implementation you've considered

Follow MySQL behavior: https://dev.mysql.com/doc/refman/8.0/en/group-by-modifiers.html

**Comment by Cyberleu** [comment]
I have read the mysql 8.0 document, can I have a try? @dengn

**Comment by YANGGMM** [comment]
ok, you can try it, thanks for your contribution @Cyberleu

**Comment by Cyberleu** [comment]
@YANGGMM Sorry, I didn't noticed that you have unassigned me. I have finished this issue, and now i'm merging conflicts.

**Comment by YANGGMM** [comment]
> @YANGGMM Sorry, I didn't noticed that you have unassigned me. I have finished this issue, and now i'm merging conflicts.

ok, I have assigned it to you, thanks for your contribution!

**Comment by YANGGMM** [comment]
@Cyberleu  do you have any interest in supporting GROUP BY WITH CUBE and GROUP BY WITH sets ?

**Comment by Cyberleu** [comment]
> @Cyberleu do you have any interest in supporting GROUP BY WITH CUBE and GROUP BY WITH sets ?
Yes,I'd like to try. And I think I will redesign 'with rollup' based on the 'grouping sets'.

**Comment by YANGGMM** [comment]
> > @Cyberleu do you have any interest in supporting GROUP BY WITH CUBE and GROUP BY WITH sets ?
> > Yes,I'd like to try. And I think I will redesign 'with rollup' based on the 'grouping sets'.

ok, will you change the pr https://github.com/matrixorigin/matrixone/pull/18678, or open a new pr?

**Comment by Cyberleu** [comment]
@YANGGMM Hi,could you please help check my pr? Thx!

**Comment by YANGGMM** [comment]
> @YANGGMM Hi,could you please help check my pr? Thx!

but there are some conflicts and test error, @Cyberleu  can you solve it first, I will  check the pr later

**Comment by Cyberleu** [comment]
> > @YANGGMM Hi,could you please help check my pr? Thx!
> 
> but there are some conflicts and test error, @Cyberleu can you solve it first, I will check the pr later

I'm testing now, there is something wrong with the privious pr, i will modify it later.

**Comment by YANGGMM** [comment]
> > > @YANGGMM Hi,could you please help check my pr? Thx!
> > 
> > 
> > but there are some conflicts and test error, @Cyberleu can you solve it first, I will check the pr later
> 
> I'm testing now, there is something wrong with the privious pr, i will modify it later.

ok, thanks

**Comment by YANGGMM** [comment]
> @YANGGMM Test done.

ok, thanks, how can I contact with you?

**Comment by Cyberleu** [comment]
> > @YANGGMM Test done.
> 
> ok, thanks, how can I contact with you?

I have set my personal profile just now.

**Comment by YANGGMM** [comment]
support
group by  with  rollup
group by  with  cube
group by  with  sets
grouping
finished, plz test @aressu1985, plz add document about it @yangj1211

**Comment by Cyberleu** [comment]
现在的语法是

**Comment by Cyberleu** [comment]
support
- group by rollup(a,b,c)
- group by cube(a,b,c)
- group by grouping sets((a,b),(b,c),())
- grouping
@aressu1985 @yangj1211

**Comment by allengaoo** [comment]
@aressu1985  this one has entered testing phase, please assign it. The recommanded release version is 2.1.0

**Comment by Ariznawlll** [comment]
After testing, the function is basically unavailable, please @YANGGMM help to find the relevant responsible person to do the relevant repair and then transfer to me, Thanks
